src/pikepdf/models/outlines.py

```diff
@@ -0,0 +1,247 @@
+        self.obj = obj
+        for k, v in kwargs.items():
+            if k in ALL_PAGE_LOCATION_KWARGS:
+                self.page_location_kwargs[k] = v
+            else:
+                raise ValueError("Invalid keyword argument {0}".format(k))
+        self.is_closed = False
+        self.children = []
+
+    def __str__(self):
+        if self.children:
+            if self.is_closed:
+                oc_indicator = '[+]'
+            else:
+                oc_indicator = '[-]'
+        else:
+            oc_indicator = '[ ]'
+        if self.destination is not None:
+            dest = self.destination
+        else:
+            dest = '<Action>'
+        return '{1} {0.title} -> {2}'.format(self, oc_indicator, dest)
+
+    def __repr__(self):
+        return '<{0.__class__.__name__}: "{0.title}">'.format(self)
+
+    @classmethod
+    def from_dictionary_object(cls, obj: Dictionary):
+        """Creates a ``OutlineItem`` from a PDF document's ``Dictionary``
+        object. Does not process nested items.
+
+        Arguments:
+            obj: ``Dictionary`` object representing a single outline node.
+        """
+        title = str(obj.Title)
+        destination = obj.get('/Dest')
+        action = obj.get('/A')
+        return cls(title, destination=destination, action=action, obj=obj)
+
+    def to_dictionary_object(self, pdf) -> Dictionary:
+        """Creates a ``Dictionary`` object from this outline node's data,
+        or updates the existing object.
+        Page numbers are resolved to a page reference on the input
+        ``Pdf`` object.
+
+        Arguments:
+            pdf: PDF document object.
+        """
+        if self.obj is None:
+            self.obj = obj = pdf.make_indirect(Dictionary())
+        else:
+            obj = self.obj
+        obj.Title = self.title
+        if self.destination is not None:
+            if isinstance(self.destination, int):
+                self.destination = get_page_destination(pdf, self.destination, self.page_location,
+                                                        **self.page_location_kwargs)
+            obj.Dest = self.destination
+            if '/A' in obj:
+                del obj.A
+        elif self.action is not None:
+            obj.A = self.action
+            if '/Dest' in obj:
+                del obj.Dest
+        return obj
+
+
+class Outline:
+    """Maintains a intuitive interface for creating and editing PDF document outlines,
+    according to the PDF reference manual (ISO32000:2008) section 12.3.
+
+    Arguments:
+        pdf: PDF document object.
+    """
+    def __init__(self, pdf):
+        self._root = []
+        self._pdf = pdf
+        if pdf:
+            self.load()
+
+    def __str__(self):
+        return str(self.root)
+
+    def __repr__(self):
+        return '<{0.__class__.__name__}: {1} items>'.format(self, len(self.root))
+
+    def _save_level_outline(self, parent: Dictionary, outline_items: list):
```

Comment:
Both _load and _save will raise a RecursionError in a PDF that contains a loop in its Outline. Such PDFs definitely exist in the wild and can be accidentally created by users when they manipulate the data structure.

Wrapping libqpdf might help, since it takes defensive measures. If not that, we will want to limit recursion depth and maintain a list of visited nodes.

Reply:
I have now implemented some checks in the PR. If we later decide to use libqpdf functionality, we can reduce or remove the Python implementation.
On both _load and _save, the tree depth is limited to 15 by default. This is a lot more than should be used in practice and can be modified by the user. _load also checks for loops by gathering all visited nodes, whereas _save checks if prior-existing Pdf object references have been reused. Both functions have a strict (failing) and a non-strict (auto-fix) mode.